userspace/libscap/scap_procs.c

```diff
@@ -700,6 +700,83 @@ int32_t scap_proc_fill_exe_writable(scap_t* handle, struct scap_threadinfo* tinf
 	return SCAP_SUCCESS;
 }
 
+void scap_proc_fill_exe_upper_layer(scap_t* handle, struct scap_threadinfo* tinfo, const char *procdirname)
+{
+	char buf[SCAP_MAX_PATH_SIZE];
+	char upperdir[SCAP_MAX_PATH_SIZE];
+	char exe[SCAP_MAX_PATH_SIZE];
+	char line[SCAP_MAX_ARGS_SIZE];
+	FILE *mounts_file;
+	struct stat upperdir_exe_stat;
+	int ret;
+
+	//
+	// Open mounts file
+	// 
+	snprintf(buf, sizeof(buf), "%smounts", procdirname);
+	mounts_file = fopen(buf, "r");
+	if(!mounts_file)
+	{
+		return;
+	}
+
+	//
+	// Check if there is an overlayfs mounted in root. 
+	// This is used to understand if the process is within a container.
+	// If this is the case, the first line of the mounts file contains the string "overlay / overlay" 
+	// and we can retrieve the upperdir path for this container.
+	//
+	fgets(line, sizeof(line), mounts_file);
+	if(strstr(line, "overlay / overlay"))
+	{
+		char *start, *end;
+
+		start = strstr(line, "upperdir=") + sizeof("upperdir=") - 1;
+		if(!start)
+		{
+			return;
```

Comment:
Seems like some cases like this do not `fclose()` the file

Reply:
Thanks for noticing this!